From libes@cme.nist.gov Thu Jan  4 11:40:54 1996
Return-Path: <libes@cme.nist.gov>
Received: from istwok.ods.com ([192.94.73.11]) by elo.ods.com
	 with smtp id <m0tXteo-000Bl6C@elo.ods.com>
	(Debian /\oo/\ Smail3.1.29.1 #29.35); Thu, 4 Jan 96 11:40 CST
Received: from muffin.cme.nist.gov by istwok.ods.com (4.1/SMI-4.1)
	id AA24141; Thu, 4 Jan 96 11:42:33 CST
Received: by muffin.cme.nist.gov (4.1/SMI-3.2-del.7)
	id AA17461; Thu, 4 Jan 96 12:42:44 EST
Date: Thu, 4 Jan 96 12:42:44 EST
From: libes@cme.nist.gov (Don Libes)
Message-Id: <9601041742.AA17461@muffin.cme.nist.gov>
To: david@elo.ods.com (David Engel)
Subject: Re: Bug#1836: expect core dumps (fwd)
In-Reply-To: <m0tXhju-000Bl7C@elo.ods.com>
References: <m0tXhju-000Bl7C@elo.ods.com>
Status: RO

David Engel writes:
>Hi Don,
>
>I maintain your expect package for the Debian distribution of Linux.
>The following bug report was sent to me some time ago.  I just retried
>it with expect 5.18.1 and the problem still exists.  I have been
>unable to find the bug so far.  Could you please look into it when you
>get a chance.  Thanks.
>
>David
>
>Forwarded message:
>> From @mongo.pixar.com:debian-devel-request@Pixar.com Fri Nov 10 01:05:42 1995
>> Old-Return-Path: <iwj10@cus.cam.ac.uk>
>> Subject: Bug#1836: expect core dumps
>> Reply-To: swift@bu.edu, debian-bugs@Pixar.com
>> Resent-From: Matthew Swift <swift@bu.edu>
>> Resent-To: debian-devel@Pixar.com
>> Resent-Date: Fri, 10 Nov 1995 07:03:02 GMT
>> Resent-Message-Id: <debian-bugs-handler.1836.B11100650560@pixar.com>
>> X-Debian-Pr-Package: expect
>> Message-Id: <199511100649.BAA01537@aleph.bu.edu>
>> To: debian-bugs@Pixar.com
>> Date: Fri, 10 Nov 1995 01:49:44 -0500
>> From: Matthew Swift <swift@bu.edu>
>> X-Mailing-List: <debian-devel@Pixar.com> archive/latest/7402
>> X-Loop: debian-devel@Pixar.com
>> Precedence: list
>> Resent-Sender: debian-devel-request@Pixar.com
>> 
>> 
>> Package: expect
>> Version: 5.16.3
>> Revision: 3
>> 
>> The following dumps core when put into a file and executed.
>> 
>> I realize I should probably spawn su inside a shell, but I don't think
>> expect should core dump even so.
>> 
>> #!/usr/bin/expect -f
>> 
>> # Give my roommate a way to safely turn off my machine without giving
>> # him the root password.
>> 
>> set password [exec cat /root/.password]
>> spawn su
>> expect Password:
>> send $password\r
>> expect #
>> send_user "Shut down completely? Are you sure? "
>> expect_user -re -nocase "y(es)?" {send_user "Closing down!"} -re -nocase "no?" {send_user "Not closing down"}
>> 
>> 
>> 
>
>
>-- 
>David Engel                        Optical Data Systems, Inc.
>david@ods.com                      1101 E. Arapaho Road
>(214) 234-6400                     Richardson, TX  75081

I think I found the problem (uninitialized variable).  I've enclosed a
patch.

After applying the patch, please try Expect on the same script that
you sent me.  Let me know how it goes.


*** expect.c.~32~	Thu Nov  2 09:59:58 1995
--- expect.c	Thu Jan  4 12:26:37 1996
***************
*** 1459,1465 ****
  		}
  		f->msize = new_msize;
  		f->buffer[f->size] = '\0';
! 
  	}
  }
  
--- 1459,1465 ----
  		}
  		f->msize = new_msize;
  		f->buffer[f->size] = '\0';
! 		f->lower[f->size] = '\0';
  	}
  }
  

After testing the patch, you'll probably want to pass the following
note back to the user who sent you the bug report:

	expect_user \
	  -re -nocase "y(es)?" {
		send_user "Closing down!"
	} -re -nocase "no?" {
		send_user "Not closing down"
	}

The "-re" flag says that "the next argument is a regular expression".
So Expect interprets your command to mean that the regular expr is
"-nocase", which is almost certainly not what you wanted.  To
interpret "-nocase" as a flag, put it before the "-re" flag:

	expect_user \
	  -nocase -re "y(es)?" {
		send_user "Closing down!"
	} -nocase -re "no?" {
		send_user "Not closing down"
	}

Don

